lld/ELF/Writer.cpp

```diff
@@ -1444,6 +1444,38 @@ static void finalizeSynthetic(Ctx &ctx, SyntheticSection *sec) {
   }
 }
 
+static bool canInsertPadding(OutputSection *sec) {
+  StringRef s = sec->name;
+  return s == ".bss" || s == ".data" || s == ".data.rel.ro" || s == ".rodata" ||
```

Comment:
This is matching on the output section name, and since the suffix on `.rodata.*` and so on is removed when forming the output section name the feature works as intended with `-fdata-sections` enabled.

The reason for using `starts_with` to match `.text` is because of sections such as `.text.hot` and `.text.unlikely` that are preserved as output section names.

There won't be a linker defined start/stop on `.data.foo` without a linker script and I think the solution there is to allow user control as noted in the main reply.

Reply:
OK. Perhaps worth adding a comment to explain the choice of sections.

Reply:
Hard coding these names is a bit too magical and we'd prefer options avoiding the magic.

Perhaps encode the output section patterns in the option argument like 
`--shuffle-sections='1;.bss;.data;.rodata;.text*'`

Reply:
To me this doesn't feel too different from the hardcoded output section name prefixes in `LinkerScript::getOutputSectionName`. I think it's important for this feature to be user friendly in the common case where there is no linker script or the linker script isn't doing anything unusual. In a future extension, we may consider adding support for custom section patterns like with `--shuffle-sections`. After this is implemented, `--shuffle-padding=1` would just become a shorthand way of writing `--shuffle-padding=.text*=1,.bss=1,...` (or whatever syntax we decide on). But this doesn't need to happen in the initial implementation of the feature.

Reply:
Should `.lrodata` `.ldata` be added?

Reply:
Makes sense, I also added `.ltext` and `.lbss`.